spec/ruby_terraform/commands/apply_spec.rb

```diff
@@ -11,280 +14,36 @@
-        receive(:spawn)
-            .with('terraform apply some/path/to/terraform/plan', any_args))
-
-    command
-        .execute(
-            directory: 'some/path/to/terraform/configuration',
-            plan: 'some/path/to/terraform/plan')
-  end
-
-  it 'defaults to the configured binary when none provided' do
-    command = RubyTerraform::Commands::Apply.new
-
-    expect(Open4).to(
-        receive(:spawn)
-            .with('path/to/binary apply some/path/to/terraform/configuration', any_args))
-
-    command.execute(directory: 'some/path/to/terraform/configuration')
-  end
-
-  it 'adds a var option for each supplied var' do
-    command = RubyTerraform::Commands::Apply.new(binary: 'terraform')
-
-    expect(Open4).to(
-        receive(:spawn)
-            .with("terraform apply -var 'first=1' -var 'second=two' some/configuration", any_args))
-
-    command.execute(
-        directory: 'some/configuration',
-        vars: {
-            first: 1,
-            second: 'two'
-        })
-  end
-
-  it 'correctly serialises list/tuple vars' do
-    command = RubyTerraform::Commands::Apply.new(binary: 'terraform')
-
-    expect(Open4).to(
-        receive(:spawn)
-            .with("terraform apply -var 'list=[1,\"two\",3]' some/configuration", any_args))
-
-    command.execute(
-        directory: 'some/configuration',
-        vars: {
-            list: [1, "two", 3]
-        })
-  end
-
-  it 'correctly serialises map/object vars' do
-    command = RubyTerraform::Commands::Apply.new(binary: 'terraform')
-
-    expect(Open4).to(
-        receive(:spawn)
-            .with("terraform apply -var 'map={\"first\":1,\"second\":\"two\"}' some/configuration", any_args))
-
-    command.execute(
-        directory: 'some/configuration',
-        vars: {
-            map: {
-                first: 1,
-                second: "two"
-            }
-        })
-  end
-
-  it 'correctly serialises vars with lists/tuples of maps/objects' do
-    command = RubyTerraform::Commands::Apply.new(binary: 'terraform')
+  terraform_command = 'apply'
 
-    expect(Open4).to(
-        receive(:spawn)
-            .with("terraform apply -var 'list_of_maps=[{\"key\":\"value\"},{\"key\":\"value\"}]' some/configuration", any_args))
+  it_behaves_like 'a command with an argument', [terraform_command, :directory]
 
-    command.execute(
-        directory: 'some/configuration',
-        vars: {
-            list_of_maps: [
-                {key: "value"},
-                {key: "value"}
-            ]
-        })
-  end
+  it_behaves_like 'a command with an argument', [terraform_command, :plan]
 
-  it 'adds a state option if a state path is provided' do
-    command = RubyTerraform::Commands::Apply.new(binary: 'terraform')
+  it_behaves_like 'a valid command line', {
+    reason: 'prefers the plan if both plan and directory provided',
+    expected_command: 'terraform apply some/path/to/terraform/plan',
+    options: { directory: DIRECTORY_PATH,
```

Comment:
I'm not sure if I like the use of global constants here. Again because of the "action at a distance" thing. I now have to go and find the constant to understand the full intent of the test / behaviour.

If I don't care about the specific value of a test variable, I'll sometimes use a `Random` module so I can say something like `Random.path`. That would avoid the global constant and express the intent of "it doesn't matter what this is as long as it looks like a path".

Something kind of like this: https://github.com/infrablocks/concourse.rb/blob/main/spec/build/data.rb

Reply:
Fair call - was working late last night and didn't really think it through too well :-)

The annoying part is, it needs to be passed in as part of the options hash and (for most of the specs) the value also forms part of the expected command line result, so it needs to be held somewhere.  I've moved it into local variables just to limit the impact.

Reply:
Great, thanks!